core/functional_unroller.cpp

```diff
@@ -0,0 +1,101 @@
+/*********************                                                        */
+/*! \file functional_unroller.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Makai Mann
+** This file is part of the pono project.
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief An unroller implementation for functional transition systems that has
+**        a configurable parameter for when to introduce new timed variables.
+**
+**
+**/
+#include "core/functional_unroller.h"
+
+#include "assert.h"
+#include "utils/exceptions.h"
+
+using namespace smt;
+using namespace std;
+
+namespace pono {
+
+FunctionalUnroller::FunctionalUnroller(const TransitionSystem & ts,
+                                       const SmtSolver & solver,
+                                       size_t interval)
+    : super(ts, solver), interval_(interval), true_(solver_->make_term(true))
+{
+  if (!ts.is_functional()) {
+    throw PonoException(
+        "Can only use FunctionalUnroller on a FunctionalTransitionSystem.");
+  }
+}
+
+Term FunctionalUnroller::at_time(const Term & t, unsigned int k)
+{
+  if (!ts_.no_next(t)) {
+    throw PonoException(
+        "Functional unroller cannot replace next state variables");
+  }
+  return super::at_time(t, k);
+}
+
+UnorderedTermMap & FunctionalUnroller::var_cache_at_time(unsigned int k)
+{
+  const UnorderedTermMap & state_updates = ts_.state_updates();
+  while (time_cache_.size() <= k) {
+    time_cache_.push_back(UnorderedTermMap());
+    extra_constraints_.push_back(true_);
+    UnorderedTermMap & subst = time_cache_.back();
+    const unsigned int t = time_cache_.size() - 1;
+    assert(extra_constraints_.size() == t + 1);
+
+    // create new state variables instead of substituting
+    // every interval_ steps (if interval_ nonzero)
+    // except when k is zero then we have to create_new regardless
+    bool create_new = (interval_ && (k % interval_ == 0));
+    create_new |= !t;
+
+    for (auto v : ts_.statevars()) {
+      if (create_new || state_updates.find(v) == state_updates.end()) {
+        Term new_v = var_at_time(v, t);
+        subst[v] = new_v;
+      }
+
+      if (t == 0) {
+        assert(create_new);  // should be creating new symbols at 0
+        // no extra constraints at 0
+        continue;
+      }
+
+      Term fun_subst =
+          solver_->substitute(state_updates.at(v), time_cache_.at(t - 1));
+
+      if (create_new) {
+        // add equality to extra constraints
+        extra_constraints_[t] =
+            solver_->make_term(And,
+                               extra_constraints_[t],
```

Comment:
where it is populated for the first time?

Reply:
It starts as just `True` here: https://github.com/upscale-project/pono/blob/ca0490aa9b7137ccfcda31a8e57e7217c2952e77/core/functional_unroller.cpp#L52